providers/coordinator.go

```diff
@@ -62,6 +63,33 @@ type RunningProvider struct {
 	lock sync.Mutex
 }
 
+// initialize the heartbeat with the provider
+func (p *RunningProvider) heartbeat() {
+	interval := 2 * time.Second
+	gracePeriod := 3 * time.Second
+	go func() {
```

Comment:
I think this needs to take some sort of context or channel to properly shut down.